Question: Http Test to bypass insecure https url

Hi @fhoeben how can we bypass an insecure https website when using http test fixture?

```bash
|script       |json http test|
| get from | https://xyz.com |
```

I'm getting Caused by: javax.net.ssl.SSLPeerUnverifiedException: Host name 'xyz.com' does not match the certificate subject provided by the peer ...

--- a/src/main/java/nl/hsac/fitnesse/fixture/Environment.java
+++ b/src/main/java/nl/hsac/fitnesse/fixture/Environment.java
@@ -781,4 +781,18 @@ public class Environment {
     public void disableHttpClientCompression() {
         httpClient.disableCompression();
     }
+
+    /**
+     * Disables SSL certificate verification on this environment's HttpClient
+     */
+    public void disableHttpClientSSLVerification() {
+        httpClient.disableSSLVerification();
+    }
+
+    /**
+     * Enables SSL certificate verification on this environment's HttpClient
+     */
+    public void enableHttpClientSSLVerification() {
+        httpClient.enableSSLVerification();
+    }
 }

--- a/src/main/java/nl/hsac/fitnesse/fixture/slim/HttpTest.java
+++ b/src/main/java/nl/hsac/fitnesse/fixture/slim/HttpTest.java
@@ -67,6 +67,20 @@ public class HttpTest extends SlimFixtureWithMap {
     }
 
     /**
+     * Disables SSL certificate verification in the current environment (i.e. for the entire test run)
+     */
+    public void disableSSLVerification() {
+        getEnvironment().disableHttpClientSSLVerification();
+    }
+
+    /**
+     * Enables SSL certificate verification in the current environment (i.e. for the entire test run)
+     */
+    public void enableSSLVerification() {
+        getEnvironment().enableHttpClientSSLVerification();
+    }
+
+    /**
      * Stores value to be passed as headers.
      * @param value value to be passed.
      * @param name name to use this value for.

--- a/src/main/java/nl/hsac/fitnesse/fixture/util/HttpClient.java
+++ b/src/main/java/nl/hsac/fitnesse/fixture/util/HttpClient.java
@@ -16,6 +16,7 @@ import org.apache.http.client.methods.HttpPost;
 import org.apache.http.client.methods.HttpPut;
 import org.apache.http.client.methods.HttpRequestBase;
 import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
 import org.apache.http.entity.ContentType;
 import org.apache.http.entity.StringEntity;
 import org.apache.http.entity.mime.MultipartEntityBuilder;
@@ -24,8 +25,10 @@ import org.apache.http.impl.client.HttpClientBuilder;
 import org.apache.http.impl.client.HttpClients;
 import org.apache.http.protocol.BasicHttpContext;
 import org.apache.http.protocol.HttpContext;
+import org.apache.http.ssl.SSLContexts;
 import org.apache.http.util.EntityUtils;
 
+import javax.net.ssl.SSLContext;
 import java.io.File;
 import java.io.IOException;
 import java.net.URI;
@@ -39,11 +42,12 @@ import java.util.Map;
 public class HttpClient {
     /** Default HttpClient instance used. */
     public final static org.apache.http.client.HttpClient DEFAULT_HTTP_CLIENT;
-    public static org.apache.http.client.HttpClient COMPRESSION_HTTP_CLIENT;
     private org.apache.http.client.HttpClient httpClient;
+    private boolean contentCompression = false,
+            sslVerification = true;
 
     static {
-        DEFAULT_HTTP_CLIENT = buildHttpClient(false);
+        DEFAULT_HTTP_CLIENT = buildHttpClient(false, true);
     }
 
     /**
@@ -161,8 +165,9 @@ public class HttpClient {
      * Ensures the apache HttpClient used supports content compression
      */
     public void enableCompression() {
-        if (httpClient == DEFAULT_HTTP_CLIENT) {
-            httpClient = getCompressionHttpClient();
+        if (!this.contentCompression) {
+            this.contentCompression = true;
+            updateHttpClient();
         }
     }
 
@@ -170,20 +175,41 @@ public class HttpClient {
      * Ensures the apache HttpClient used does not support content compression
      */
     public void disableCompression() {
-        if (httpClient != DEFAULT_HTTP_CLIENT) {
-            httpClient = DEFAULT_HTTP_CLIENT;
+        if (this.contentCompression) {
+            this.contentCompression = false;
+            updateHttpClient();
         }
     }
 
     /**
-     * Returns an apache HttpClient with content compression support (builds it only on first call)
-     * @return an apache HttpClient with content compression support
+     * Ensures the apache HttpClient used does not verify SSL certificates
      */
-    protected org.apache.http.client.HttpClient getCompressionHttpClient() {
-        if (COMPRESSION_HTTP_CLIENT == null) {
-            COMPRESSION_HTTP_CLIENT = buildHttpClient(true);
+    public void disableSSLVerification() {
+        if (this.sslVerification) {
+            this.sslVerification = false;
+            updateHttpClient();
         }
-        return COMPRESSION_HTTP_CLIENT;
+    }
+
+    /**
+     * Ensures the apache HttpClient used verifies SSL certificates
+     */
+    public void enableSSLVerification() {
+        if (!this.sslVerification) {
+            this.sslVerification = true;
+            updateHttpClient();
+        }
+    }
+
+    /**
+     * Sets the apache HttpClient to one matching the current contentCompression and sslVerification values
+     */
+    protected void updateHttpClient() {
+        if (!contentCompression && sslVerification) {
+            this.httpClient = DEFAULT_HTTP_CLIENT;
+        } else {
+            this.httpClient = buildHttpClient(this.contentCompression, this.sslVerification);
+        }
     }
 
     /**
@@ -191,7 +217,7 @@ public class HttpClient {
      * @param contentCompression if true, the returned instance will support content compression
      * @return an apache HttpClient instance
      */
-    protected static org.apache.http.client.HttpClient buildHttpClient(boolean contentCompression) {
+    protected static org.apache.http.client.HttpClient buildHttpClient(boolean contentCompression, boolean sslVerification) {
         RequestConfig rc = RequestConfig.custom()
                 .setCookieSpec(CookieSpecs.STANDARD)
                 .build();
@@ -203,10 +229,26 @@ public class HttpClient {
             builder.disableContentCompression();
         }
 
+        if (!sslVerification) {
+            try {
+                builder.setSSLSocketFactory(generateAllTrustingSSLConnectionSocketFactory());
+            } catch (Exception e) {
+                throw new RuntimeException("Unable to create all-trusting SSLConnectionSocketFactory", e);
+            }
+
+        }
+
         return builder.setConnectionReuseStrategy(NoConnectionReuseStrategy.INSTANCE)
                 .setUserAgent(HttpClient.class.getName())
                 .setDefaultRequestConfig(rc)
                 .build();
+    }
+
+    protected static SSLConnectionSocketFactory generateAllTrustingSSLConnectionSocketFactory() throws Exception {
+        SSLContext allTrustingSSLContext = SSLContexts.custom()
+                .loadTrustMaterial(null, (a, b) -> true)
+                .build();
+        return new SSLConnectionSocketFactory(allTrustingSSLContext, (a, b) -> true);
     }
 
     protected void getResponse(String url, HttpResponse response, HttpRequestBase method, Map<String, Object> headers) {

--- a/wiki/FitNesseRoot/HsacExamples/SlimTests/HttpTests.wiki
+++ b/wiki/FitNesseRoot/HsacExamples/SlimTests/HttpTests.wiki
@@ -16,6 +16,7 @@ The >HttpPost2UsingScenarioTest makes multiple SOAP calls to a service, where sc
 >HttpCookieHandling shows how cookie handling can be configured.
 >HttpHeadTest shows an example HEAD request, retrieving only response status and headers.
 >HttpCompressionTest shows an example GET request with gzip compression support.
+>HttpsWithoutCertificateVerification shows example GET requests without SSL certificate verification
 
 !2 Language
 
@@ -77,6 +78,8 @@ The main commands/keywords (i.e. public methods) offered by !-nl.hsac.fitnesse.f
 |repeat interval                                                       |Returns the number of milliseconds between requests sent by 'repeat until ...' commands.                                                                                                |
 |enable compression                                                    |Enables content compression support (e.g. gzip) for all subsequent requests until the end of the test run                                                                               |
 |disable compression                                                   |Disables content compression support for all subsequent requests until the end of the test run (this is the default state)                                                              |
+|disable SSL verification                                              |Use at your own risk! Disables SSL certificate verification for all subsequent requests until the end of the test run. This makes https potentially as unsafe as http!                  |
+|enable SSL verification                                               |Enables SSL certificate verification for all subsequent requests until the end of the test run (this is the default state)                                                              |
 
 !3 !-JsonHttpTest-!
 The commands/keywords (i.e. public methods) added by !-nl.hsac.fitnesse.fixture.slim.JsonHttpTest-! are listed below.

--- a/wiki/FitNesseRoot/HsacExamples/SlimTests/HttpTests/HttpsWithoutCertificateVerification.wiki
+++ b/wiki/FitNesseRoot/HsacExamples/SlimTests/HttpTests/HttpsWithoutCertificateVerification.wiki
@@ -1,0 +1,20 @@
+---
+Test
+---
+!2 Https without certificate verification
+
+Performs HTTPS GET on domains that would normally throw an exception because of bad SSL certificates.
+
+'''''Use at your own risk, this makes https potentially as unsafe as http!'''''
+
+|script  |http test                         |
+|disable SSL verification                   |
+|get from|https://expired.badssl.com/       |
+|get from|https://wrong.host.badssl.com/    |
+|get from|https://self-signed.badssl.com/   |
+|get from|https://untrusted-root.badssl.com/|
+
+Note that 'disable SSL verification' lasts for the rest of the test run, or until you call 'enable SSL verification', like this:
+
+|script    |http test   |
+|enable SSL verification|

